Enable Parallel test execution in pkg/scheduler

electrocucaracha opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind feature

<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Since go v1.7, the go testing package provides the ability to run [tests in parallel](https://pkg.go.dev/testing#T.Parallel). As this [post](https://engineering.mercari.com/en/blog/entry/20220408-how_to_use_t_parallel/) describes, unit tests will run in parallel at the package level. Therefore, this change allows increasing the number of processes during the execution of SIG/scheduling unit tests

* Before (forcing testing cleanup):
```bash
$ for i in {1..3}; do make test WHAT=./pkg/scheduler/; done
+++ [0914 12:42:23] Building go targets for linux/amd64
    k8s.io/kubernetes/hack/make-rules/helpers/go2make (non-static)
+++ [0914 12:42:25] Running tests without code coverage and with -race
ok      k8s.io/kubernetes/pkg/scheduler 6.445s
+++ [0914 12:42:34] Building go targets for linux/amd64
    k8s.io/kubernetes/hack/make-rules/helpers/go2make (non-static)
+++ [0914 12:42:37] Running tests without code coverage and with -race
ok      k8s.io/kubernetes/pkg/scheduler 6.424s
+++ [0914 12:42:46] Building go targets for linux/amd64
    k8s.io/kubernetes/hack/make-rules/helpers/go2make (non-static)
+++ [0914 12:42:48] Running tests without code coverage and with -race
ok      k8s.io/kubernetes/pkg/scheduler 6.461s
```
* After (forcing testing cleanup):
```bash
$ for i in {1..3}; do make test WHAT=./pkg/scheduler/; done
+++ [0914 12:43:59] Building go targets for linux/amd64
    k8s.io/kubernetes/hack/make-rules/helpers/go2make (non-static)
+++ [0914 12:44:01] Running tests without code coverage and with -race
ok      k8s.io/kubernetes/pkg/scheduler 1.160s
+++ [0914 12:44:05] Building go targets for linux/amd64
    k8s.io/kubernetes/hack/make-rules/helpers/go2make (non-static)
+++ [0914 12:44:08] Running tests without code coverage and with -race
ok      k8s.io/kubernetes/pkg/scheduler 1.160s
+++ [0914 12:44:12] Building go targets for linux/amd64
    k8s.io/kubernetes/hack/make-rules/helpers/go2make (non-static)
+++ [0914 12:44:14] Running tests without code coverage and with -race
ok      k8s.io/kubernetes/pkg/scheduler 1.159s
```



#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
NA

#### Special notes for your reviewer:

I have to force the testing cache to make sure improvements are achieved
```bash
$ git diff hack/make-rules/test.sh
diff --git a/hack/make-rules/test.sh b/hack/make-rules/test.sh
index d58bc929752..01465b9be7b 100755
--- a/hack/make-rules/test.sh
+++ b/hack/make-rules/test.sh
@@ -257,6 +257,7 @@ runTests() {
 
   verifyPathsToPackagesUnderTest "$@"
 
+  go clean -testcache
   # If we're not collecting coverage, run all requested tests with one 'go test'
   # command, which is much faster.
   if [[ ! ${KUBE_COVER} =~ ^[yY]$ ]]; then
```

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

electrocucaracha commented:
Related to #112268

chendave reviewed (commented):
Just a question, what's the difference with test option "test.parallel"? 
I am fine with adding this to speed up the test execution, but this will force to check all those testcases instead of just eventhandlers.

electrocucaracha commented:
> Just a question, what's the difference with test option "test.parallel"? I am fine with adding this to speed up the test execution, but this will force to check all those testcases instead of just eventhandlers.

[AFAIK](https://stackoverflow.com/a/45974915/2727227), the `-parallel` and `-test.parallel` flags controls the maximum number of tests to run simultaneously. The `t.Parallel()` method signals that test to be run in parallel with another parallel tests.

aojea commented:
I'm not sure we should do this, there are some tests that depend on global variables, the ones that use metrics per example ... you can check a list of test that can not run in parallel here, this job runs the same test 10 times, and some of them fail because depend on some global state https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-generate-make-test-count10

I'm -1 on implementing this until:
- we have a clear indication of the improvement, i.e. reduce CI execution in X%, 
- we have those tests green on https://testgrid.k8s.io/sig-testing-canaries#ci-kubernetes-generate-make-test-count10

electrocucaracha commented:
@aojea I agree with you, this needs to be a better analysis or take a look case by case. Unfortunately, in order to see improvements, a majority of test cases needs to implement this parallelism (kind of chicken and egg situation). Anyway, we can consider this approach for future testing implementations, and leave the existing ones for further analysis.

alculquicondor commented:
The tests touched in this PR seem fine to run in parallel.

Do you have proof of improvements in time execution?

electrocucaracha commented:
> The tests touched in this PR seem fine to run in parallel.
> 
> Do you have proof of improvements in time execution?

Maybe I need to do better analysis because my results are giving me similar performance numbers.

```shell
$ make clean
+++ [0908 12:34:02] Verifying Prerequisites....             
+++ [0908 12:34:03] Removing _output directory
$ GOMAXPROCS=2 time make test WHAT=./pkg/scheduler/
+++ [0908 12:34:07] Building go targets for linux/amd64     
    k8s.io/kubernetes/hack/make-rules/helpers/go2make (non-static)
+++ [0908 12:34:17] Building go targets for linux/amd64  
    k8s.io/code-generator/cmd/prerelease-lifecycle-gen (non-static)
+++ [0908 12:34:21] Generating prerelease lifecycle code for 27 targets
+++ [0908 12:34:23] Building go targets for linux/amd64
    k8s.io/code-generator/cmd/deepcopy-gen (non-static)
+++ [0908 12:34:26] Generating deepcopy code for 242 targets
+++ [0908 12:34:31] Building go targets for linux/amd64      
    k8s.io/code-generator/cmd/defaulter-gen (non-static)
+++ [0908 12:34:32] Generating defaulter code for 95 targets   
+++ [0908 12:34:39] Building go targets for linux/amd64               
    k8s.io/code-generator/cmd/conversion-gen (non-static)
+++ [0908 12:34:42] Generating conversion code for 132 targets                                                                                               
+++ [0908 12:34:56] Building go targets for linux/amd64           
    k8s.io/kube-openapi/cmd/openapi-gen (non-static)      
+++ [0908 12:35:08] Generating openapi code for KUBE
+++ [0908 12:35:27] Generating openapi code for AGGREGATOR
+++ [0908 12:35:28] Generating openapi code for APIEXTENSIONS
+++ [0908 12:35:29] Generating openapi code for CODEGEN
+++ [0908 12:35:30] Generating openapi code for SAMPLEAPISERVER
+++ [0908 12:35:31] Running tests without code coverage and with -race
ok      k8s.io/kubernetes/pkg/scheduler 6.467s
277.09user 27.85system 2:46.29elapsed 183%CPU (0avgtext+0avgdata 939796maxresident)k
0inputs+3395216outputs (24148major+14145651minor)pagefaults 0swaps
$ git checkout add_tparallel-scheduling
Switched to branch 'add_tparallel-scheduling'
$ make clean           
+++ [0908 12:39:27] Verifying Prerequisites....
+++ [0908 12:39:27] Removing _output directory
$ GOMAXPROCS=2 time make test WHAT=./pkg/scheduler/
+++ [0908 12:39:35] Building go targets for linux/amd64
    k8s.io/kubernetes/hack/make-rules/helpers/go2make (non-static)
+++ [0908 12:39:44] Building go targets for linux/amd64
    k8s.io/code-generator/cmd/prerelease-lifecycle-gen (non-static)
+++ [0908 12:39:49] Generating prerelease lifecycle code for 27 targets
+++ [0908 12:39:51] Building go targets for linux/amd64
    k8s.io/code-generator/cmd/deepcopy-gen (non-static)
+++ [0908 12:39:53] Generating deepcopy code for 242 targets
+++ [0908 12:39:58] Building go targets for linux/amd64
    k8s.io/code-generator/cmd/defaulter-gen (non-static)
+++ [0908 12:40:00] Generating defaulter code for 95 targets
+++ [0908 12:40:07] Building go targets for linux/amd64
    k8s.io/code-generator/cmd/conversion-gen (non-static)
+++ [0908 12:40:09] Generating conversion code for 132 targets
+++ [0908 12:40:23] Building go targets for linux/amd64
    k8s.io/kube-openapi/cmd/openapi-gen (non-static)
+++ [0908 12:40:35] Generating openapi code for KUBE
+++ [0908 12:40:53] Generating openapi code for AGGREGATOR
+++ [0908 12:40:54] Generating openapi code for APIEXTENSIONS
+++ [0908 12:40:56] Generating openapi code for CODEGEN
+++ [0908 12:40:57] Generating openapi code for SAMPLEAPISERVER
+++ [0908 12:40:58] Running tests without code coverage and with -race
ok      k8s.io/kubernetes/pkg/scheduler 6.273s
281.89user 27.54system 2:47.10elapsed 185%CPU (0avgtext+0avgdata 955840maxresident)k
0inputs+3395216outputs (24155major+14169833minor)pagefaults 0swaps
```

chendave commented:
> Maybe I need to do better analysis because the results that I'm getting are similar.

since you just update the eventhandler, so if you want to check the improvement, you should focus on eventhandler's testcases, I tried to run this with `schedule_one_test.go`, it does improve one second.

chendave commented:
On master:

go test k8s.io/kubernetes/pkg/scheduler
ok      k8s.io/kubernetes/pkg/scheduler 6.115s


With this change, 
go test k8s.io/kubernetes/pkg/scheduler
ok      k8s.io/kubernetes/pkg/scheduler 5.914s

note that I removed the test of `extender_test.go` and `scheduler_test.go`.

electrocucaracha commented:
Makes sense @chendave , I'm going to try one by one. BTW, I'm trying to reduce the issues reported by [this linter](https://github.com/kunwardeep/paralleltest)

electrocucaracha commented:
Ok, I'll break down this PR in changes per file to manage it better

electrocucaracha committed:
Enable paralellism in scheduler unit tests

Unit tests run in parallel at the package level. This change allows the
execution of sig/scheduling unit tests in parallel.

Signed-off-by: Victor Morales <v.morales@samsung.com>

electrocucaracha commented:
@aojea @alculquicondor @chendave According to my metrics, these changes reduces ~5.5x the testing time in the tests declared under pkg/scheduler folder. I couldn't apply parallelism in the following sections:

* pkg/scheduler/extender_test.go:274
* pkg/scheduler/schedule_one_test.go:545
* pkg/scheduler/schedule_one_test.go:954
* pkg/scheduler/schedule_one_test.go:2074
* pkg/scheduler/schedule_one_test.go:2113
* pkg/scheduler/schedule_one_test.go:2167
* pkg/scheduler/schedule_one_test.go:2185
* pkg/scheduler/schedule_one_test.go:2441

alculquicondor commented:
can you share some numbers?

electrocucaracha commented:
> can you share some numbers?

@alculquicondor I included them in the PR description, but basically they went from ~6.443 sec to ~1.16 sec.

alculquicondor commented:
Thanks!
@chendave can you take a first pass?

aojea commented:
> these changes reduces ~5.5x the testing time in the tests declared under pkg/scheduler folder.

now we are talking :smile: :muscle:

aojea commented:
the verify job failure is legit, you have some gofmt errors

aojea commented on pkg/scheduler/extender_test.go:
```diff
@@ -488,7 +496,9 @@ func TestConvertToVictims(t *testing.T) {
 		},
 	}
 	for _, tt := range tests {
+		tt:=tt
```
```suggestion
		tt :=  tt
```

there has to be an space after

electrocucaracha replied:
Oops, I forgot to fix the format

chendave commented on pkg/scheduler/eventhandlers_test.go:
```diff
@@ -70,7 +71,9 @@ func TestNodeAllocatableChanged(t *testing.T) {
 	} {
+		test := test
 		t.Run(test.Name, func(t *testing.T) {
+			t.Parallel()
```
what's the difference if we just keep this line and remove line 51? same question for other occurrences.

electrocucaracha replied:
Apparently this is required to run the Subtests in Parallel. This [post](https://gist.github.com/kunwardeep/80c2e9f3d3256c894898bae82d9f75d0) provides a full explanation of that.

chendave replied:
what I meant is now that all sub-tests are run in parallel, do we still need parallel the parent test? 

If this is the case, you might consider to add this line to other test for consistency, e.g. `TestFindFitSomeError`

electrocucaracha replied:
Unfortunately, we still need to call the Parallel method from the parent test. Regarding the inconsistency, during my tests I found some test method doesn't support parallelism yet or at least I got runtime errors when I tried to enable it.

alculquicondor replied:
```
func TestA(t *testing.T) {
  t.Parallel()
}
func TestB(t *testing.T) {
  t.Parallel()
}
```

This means that `TestA` and `TestB` can run in parallel. That's why you see a difference in runtime if you remove the "parent" Parallel.

chendave commented:
verified by myself, 

- on master

```
+++ [0921 17:24:48] Running tests without code coverage and with -race
ok      k8s.io/kubernetes/pkg/scheduler 1.213s
```

-  with this change

```
+++ [0921 17:33:38] Running tests without code coverage and with -race
ok      k8s.io/kubernetes/pkg/scheduler 6.365s
```

- with this change but no parallel in the parent test (https://gist.github.com/chendave/419aede0de758e25df20d77fe3985c95), 
```
+++ [0921 17:56:15] Running tests without code coverage and with -race
ok      k8s.io/kubernetes/pkg/scheduler 1.881s
```

I am fine to parallel both the parent and sub-tests even though I cannot fully understand the logic behind but it's not a big deal.


/lgtm

/assign @alculquicondor @aojea

alculquicondor commented:
/retitle Enable Parallel test execution in pkg/scheduler

alculquicondor commented:
/release-note-none
/approve

electrocucaracha commented:
> 
> I am fine to parallel both the parent and sub-tests even though I cannot fully understand the logic behind but it's not a big deal.
> 

@chendave I hope the [README](https://github.com/kunwardeep/paralleltest/blob/main/README.md) file of the linter helps to clarify more of this.

enj commented:
I have opened #112702 to revert this change as it is causing CI to flake on data races.
